### DISCUSSION TITLE
explicit m:n relationships incorrect handling of disconnect

### DISCUSSION THREAD

**Discussion by mwillbanks** [prisma/prisma#15713] [Q&A] [answered]
### Bug description

M:N relationships are special in that they have two connecting sides, on a disconnect of either one of them, internally prisma should be deleting the record. However, this does not appear to be the case but rather you receive an error message: _"nThe change you are trying to make would violate the required relation 'MToN' between the `M` and `MtoN` models"_

### How to reproduce

```
generator client {
  provider        = "prisma-client-js"
  previewFeatures = ["referentialIntegrity"]
}


model M {
  id String @id @default(cuid())
  ns MToN[]
}

model N {
  id String @id @default(cuid())
  ms MtoN[]
}

model MtoN {
  id  String @id @default(cuid())
  mId String
  nId String
  ms  M @relation(fields: [mId], references: [id], onUpdate: Cascade, onDelete: Cascade)
  ns  N @relation(fields: [nId], references: [id], onUpdate: Cascade, onDelete: Cascade)
}
```

Add a few records into the tables and join them together, grab a record from M and disconnect from N.

```
await prisma.M.update({
  data: {
    ns: {
      disconnect: [{
        id: "XXXXXXXXXXX"
      }]
    }
  }
});
```

💣 🔴 *The change you are trying to make would violate the required relation 'MToN' between the `M` and `MToN` models*

### Expected behavior

In *explicit* m:n relationships, a disconnect should internally run a delete rather than attempting to set the value to null.

### Prisma information

See reproduction case.

### Environment & setup

- OS: macOS, Debian
- Database: PostgreSQL 14.4
- Node.js version: v16.17.1

### Prisma Version

```
prisma                  : 4.3.1
@prisma/client          : 4.3.1
Current platform        : darwin
Query Engine (Node-API) : libquery-engine c875e43600dfe042452e0b868f7a48b817b9640b (at node_modules/@prisma/engines/libquery_engine-darwin.dylib.node)
Migration Engine        : migration-engine-cli c875e43600dfe042452e0b868f7a48b817b9640b (at node_modules/@prisma/engines/migration-engine-darwin)
Introspection Engine    : introspection-core c875e43600dfe042452e0b868f7a48b817b9640b (at node_modules/@prisma/engines/introspection-engine-darwin)
Format Binary           : prisma-fmt c875e43600dfe042452e0b868f7a48b817b9640b (at node_modules/@prisma/engines/prisma-fmt-darwin)
Format Wasm             : @prisma/prisma-fmt-wasm 4.3.0-32.c875e43600dfe042452e0b868f7a48b817b9640b
Default Engines Hash    : c875e43600dfe042452e0b868f7a48b817b9640b
Studio                  : 0.473.0
Preview Features        : referentialIntegrity
```

**Comment by mwillbanks** [comment]
I should note, that if you make the relationships optional, disconnect will set them to null thus leaving behind orphan records which is a _somewhat_ workaround but then requires to clean up the database when this is encountered because one side of the m:n relationship will believe that it has a correct pairing.

**Comment by janpio** [comment]
For Prisma you do not have a m:n here, but two 1:n relations. When you call `disconnect` in one of these, it tries to do exactly what you describe but in this case is blocked because of the other 1:n relation. If you want to delete relations here, you need to handle this complexity yourself unfortunately.

What you want should work with _implicit m:n relations_: https://www.prisma.io/docs/concepts/components/prisma-schema/relations/many-to-many-relations#implicit-many-to-many-relations There the PRisma schema explicitly has a m:n relation between two sides, and then can handle all the things automatically. (Downside: You have no control about the join table/model)

**Comment by mwillbanks** [comment, marked as answer]
@janpio totally understand that since the prisma client forces the 1:n relations in that way if we want additional columns on the m:n type of table. In this case, you are correct in that it is not necessarily a true m:n anymore, however, from the prisma documentation standpoint it is what it ultimately is.  Unfortunately for me, I had to add a primary key column in there due to another tools incompatibility with a multi-primary key 🤷.

it would be nice if prisma was able to handle this portion of complexity, although, it might be possible for me to define this type of action through middleware in order to handle it automatically. I suppose the real feature request would be to have true m:n relations rather than the implicit many-to-many.

**Reply by janpio** [reply]
I think that should cover this request: https://github.com/prisma/prisma/issues/2186

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
